\documentclass[a4paper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\inputencoding{utf8}
\item[Type]\AsIs{Package}
\item[Title]\AsIs{What the Package Does (Title Case)}
\item[Version]\AsIs{0.1.0}
\item[Author]\AsIs{Yujiao}
\item[Maintainer]\AsIs{The package maintainer }\email{yourself@somewhere.net}\AsIs{}
\item[Description]\AsIs{More about what it does (maybe more than one line)
Use four spaces when indenting paragraphs within the Description.}
\item[License]\AsIs{What license is it under?}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[RoxygenNote]\AsIs{6.1.0}
\item[Imports]\AsIs{BSgenome,
Biostrings,
GenomicRanges,
IRanges,
rlist,
tibble,
TnT}
\item[Suggests]\AsIs{knitr,
rmarkdown}
\item[VignetteBuilder]\AsIs{knitr}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{Class\_frame}{Class\_frame}{Class.Rul.frame}
%
\begin{Description}\relax
Based on dataframe converted by Convert\_seq funciton, this function can
divide the target seqence into Hotspot and Exome two list by thier sequence length
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Class_frame(dataframe.seq, n)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dataframe.seq}] is the dataframe converted by Convert\_seq funciton

\item[\code{n}] is the lenght of sequence, if the target sequecne is less than
n, it will belong to Hotspot, otherwise is Exome
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
Chr<-c("chrX","chrX" ,"chrX","chrX" )
Target.Start<-c(153996577,153999034,154004462,154002877)
Target.End<-c(153996707 ,153999154,154004599,154002980)
Strand<-rep("+",4)
chromosome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
frame<-Convert_seq(Chr,Target.Start,Target.End,Strand,chromosome)
class<-Class_frame(frame,125)
sapply(class,nrow)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Convert\_seq}{Convert\_seq}{Convert.Rul.seq}
%
\begin{Description}\relax
You can extract the targe sequence and convert it into dataframe
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Convert_seq(Chr, Target.Start, Target.End, Strand, chromosome, NO. = NA)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{Chr}] Chromosome information

\item[\code{Target.Start}] Starting location of target sequence

\item[\code{Target.End}] ending ocation of target sequence

\item[\code{Strand}] +/- strand

\item[\code{chromosome}] chromosome information
e.g chromosome <- BSgenome.Hsapiens.UCSC.hg19

\item[\code{NO.}] name of the each target sequence;default is NULL
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
Chr<-c("chrX","chrX" ,"chrX","chrX" )
Target.Start<-c(153996577,153999034,154004462,154002877)
Target.End<-c(153996707 ,153999154,154004599,154002980)
Strand<-rep("+",4)
chromosome <- BSgenome.Hsapiens.UCSC.hg19
Convert_seq(Chr,Target.Start,Target.End,Strand,chromosome)

\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Exo\_extension}{Exo\_extension}{Exo.Rul.extension}
%
\begin{Description}\relax
For the target sequecne dataframe, if you  need to shear, you can directly use Exo\_extension function,
which not only shears the target sequence,but also amplifies the fragments for designing the primer. Finally,
sequence of sheared fragment and amplified fragment will be returned to the original dataframe, respectively
called Fragment.seq and Extend.fragment. Meanwhile, it additionally show the location and length information of sheared
fragments. Of course，each of complete amplifyed fragment sequence will also showed at dataframe, called Extend.seq.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exo_extension(EXOME, n, cut.length, chromosome)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{EXOME}] is dataframe converted by Convert\_seq funciton and need to be sheared

\item[\code{n}] total lenght of target sequence after amplifying

\item[\code{cut.length}] Clip length of target sequecne, the sheared fragment is reuqired to be less than "cut.length"
\#e.g. for 345bp length of target sequence, it can be sheared into 1-178 and 179-354 two fragments

\item[\code{chromosome}] is the chromosome information for amplified fragment
e.g chromosome <- BSgenome.Hsapiens.UCSC.hg19
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
Chr<-c("chrX","chrX" ,"chrX","chrX" )
Target.Start<-c(153996577,153999034,154004462,154002877)
Target.End<-c(153996707 ,153999154,154004599,154002980)
Strand<-rep("+",4)
chromosome <- BSgenome.Hsapiens.UCSC.hg19：：BSgenome.Hsapiens.UCSC.hg19
frame<-Convert_seq(Chr,Target.Start,Target .End,Strand,chromosome)
class<-Class_frame(frame,125)
sapply(class,nrow)
chromosome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
entend.exome<-Exo_extension(class$EXOME,280,200,chromosome)
View(entend.exome)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Exo\_output}{Exo\_output}{Exo.Rul.output}
%
\begin{Description}\relax
This function can write the filtered result exome sequecne into '.CSV' formation. Here, it only chooses the optimal
priemr pairs for each target sequence. If there is not primer, it only returns NA result; if  no
primer pairs fit the condiditons, it returns "None of them fit" at the first column
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exo_output(fil.exo, exo.ex, NAME, Write.out = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fil.exo}] is the final filtered result of exome sequecne, namely non-sheared sequence

\item[\code{exo.ex}] is the exome targe sequence dataframe after extension processed by
'Exo\_extension' funciton

\item[\code{NAME}] name of put file
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
sp.ex<-Split_exome(exo.out,frame=exo)
exomepair.set<-List_PrimerExomeSet(sp.ex) %>% List_PairExomeSet()
fil.exo<-Filter_InfExome(exomepair.set,eplist,distance=20,overlap=10)
entend.exome<-Exo_extension(exo,280,200,chromosome)
Exo_output(fil.exo, entend.exome$Extend.seq,Write.out=FALSE)

Output:
Exo_output(fil.exo, entend.exome$Extend.seq,"exoResult")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Filter\_Inf}{Filter\_Inf}{Filter.Rul.Inf}
%
\begin{Description}\relax
Filter\_Inf not only filter the primer pairs, but also extract the basical information for each primer
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Filter_Inf(hot.pair, plist, distance, overlap, inf = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{hot.pair}] is the primer pair

\item[\code{plist}] is the list formation of  Primers under each target sequence

\item[\code{distance}] threshold for distance length between primer pairs, the distacne between primer pairs reqiure
to be less than the threshold

\item[\code{overlap}] threshold for ovelaps length between primer pairs,the overlap between primer pairs reqiure
to be less than the threshold
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
plist<-List_PrimerSet(hot.out)
hot.pair<-List_PairSet(plist)
Filter_Inf(hot.pair,plist,distance=20,overlap=15)

#Don't return information of primer
Filter_Inf(hot.pair,plist,distance=20,overlap=15,inf=FALSE)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Filter\_InfExome}{Filter\_InfExome}{Filter.Rul.InfExome}
%
\begin{Description}\relax
Filter\_InfExome is specifically applied in process of Exome sequence.
it not only filter the primer pairs, but also extract the basical information for each primer
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Filter_InfExome(exomepair.set, eplist, distance, overlap, inf = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{exomepair.set}] is the list formation of  Primers under each Exome sheared fragment

\item[\code{eplist}] list that integrates sheared fragements and corresponding primers

\item[\code{distance}] threshold for distance length between primer pairs, the distacne between primer pairs reqiure
to be less than the threshold

\item[\code{overlap}] threshold for ovelaps length between primer pairs,the overlap between primer pairs reqiure
to be less than the threshold
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
sp.ex<-Split_exome(exo.out,frame=exo)
eplist<-List_PrimerExomeSet(sp.ex)
exomepair.set<-List_PairExomeSet(eplist)
Filter_InfExome(exomepair.set,eplist,distance=20,overlap=10)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Hot\_extension}{Hot\_extension}{Hot.Rul.extension}
%
\begin{Description}\relax
For the target sequecne dataframe, if you do not need to shear, you can directly use it to
extend the target sequence for designing the primer. Finally, amplified fragment will return
in the target sequecne dataframe, called Extend.fragment. Meanwhile, it additionally show the complete
sequence after amplifying, called Extend.seq
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Hot_extension(HOTSPOT, n, chromosome)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{HOTSPOT}] is dataframe converted by Convert\_seq funciton

\item[\code{n}] total lenght of target sequence after amplifying

\item[\code{chromosome}] is the chromosome information for amplified fragment
e.g chromosome <- BSgenome.Hsapiens.UCSC.hg19
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.fame
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
Chr<-c("chrX","chrX" ,"chrX","chrX" )
Target.Start<-c(153996577,153999034,154004462,154002877)
Target.End<-c(153996707 ,153999154,154004599,154002980)
Strand<-rep("+",4)
chromosome <- BSgenome.Hsapiens.UCSC.hg19
frame<-Convert_seq(Chr,Target.Start,Target.End,Strand,chromosome)
class<-Class_frame(frame,125)
sapply(class,nrow)
chromosome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
entend.hot<-Hot_extension(class$HOTSPOT,185,chromosome)
View(entend.hot)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Hot\_output}{Hot\_output}{Hot.Rul.output}
%
\begin{Description}\relax
This function can write the filtered result into '.CSV' formation. Here, it only chooses the optimal
priemr pairs for each target sequence. If there is not primer, it only returns NA result; if  no
primer pairs fit the condiditons, it returns "None of them fit" at the first column
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Hot_output(fil.hot, NAME, Write.out = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fil.hot}] is the final filtered result of hotspot sequecne, namely non-sheared sequence

\item[\code{NAME}] name of put file
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
plist<-List_PrimerSet(hot.out)
hot.pair<-List_PairSet(plist)
rem_Pairs.inf<-Filter_Inf(hot.pair,plist,distance=20,overlap=10)
Hot_output(rem_Pairs.inf,Write.out=FALSE)

Output:
Hot_output(rem_Pairs.inf,"hotResult")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{List\_PairExomeSet}{List\_PairExomeSet}{List.Rul.PairExomeSet}
%
\begin{Description}\relax
This functions is specifically used for matching primer pair under each sheared fragement of Exome sequene and
the high quality primers are ranked at top
\end{Description}
%
\begin{Usage}
\begin{verbatim}
List_PairExomeSet(eplist, dir = "left")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{eplist}] is the list containing Exome sheared fragments and  corresponding primers achieved by 'List\_PrimerExomeSet' function

\item[\code{dir}] is the direction of extension of target sequence, if the strand of target sequence
is "+", the direction of extension of target sequence will be "left"; The direction of
"-" strand is "right". sort order of GSP1 and GSP2 depends on dir' parameter. Here, default is "left".
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
library(xxx)
data(exo.out)
data(exo)
sp.ex<-Split_exome(exo.out,frame=exo)
eplist<-List_PrimerExomeSet(sp.ex)
List_PairExomeSet(eplist)
List_PairExomeSet(eplist,dir="right")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{List\_PairSet}{List\_PairSet}{List.Rul.PairSet}
%
\begin{Description}\relax
List\_PairSet can match primer pair, the high quality primers are ranked at top
\end{Description}
%
\begin{Usage}
\begin{verbatim}
List_PairSet(plist, dir = "left")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{plist}] is the list formation of  Primers under each target sequence.Thereinto,
GSP1 is the gene specifical primer1, which is target primer and far from extension
sequence；the other closer primer is GSP2.

\item[\code{dir}] is the direction of extension of target sequence, if the strand of target sequence
is "+", the direction of extension of target sequence will be "left"; The direction of
"-" strand is "right". sort order of GSP1 and GSP2 depends on dir' parameter. Here, default is "left".
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
plist<-List_PrimerSet(hot.out)
List_PairSet(plist)
List_PairSet(plist,dir="right")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{List\_PrimerExomeSet}{List\_PrimerExomeSet}{List.Rul.PrimerExomeSet}
%
\begin{Description}\relax
This function is used to integrate sheared fragements and corresponding primers into list formation
\end{Description}
%
\begin{Usage}
\begin{verbatim}
List_PrimerExomeSet(split.exome)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{split.exome}] is the list converted from Exome Primer3 result by 'Spliy\_exome' function
Thereinto, GSP1 is the gene specifical primer1, which is target primer and far from extension
sequence；the other closer primer is GSP2
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
sp.ex<-Split_exome(exo.out,frame=exo)
x<-List_PrimerExomeSet(sp.ex)
sapply(x,function(x){c(names(x[1]),names(x[[1]]))})
only for one target
y<-List_PrimerSet(sp.ex[[1]])
sapply(y,names)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{List\_PrimerSet}{List\_PrimerSet}{List.Rul.PrimerSet}
%
\begin{Description}\relax
List\_PrimerSet
\end{Description}
%
\begin{Usage}
\begin{verbatim}
List_PrimerSet(x)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] is the output result of Primer3
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
list<-List_PrimerSet(hot.out)
sapply(list,names)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{param\_primer3}{param\_primer3}{param.Rul.primer3}
%
\begin{Description}\relax
Due to there are many parameters required by Primer3, you can specifically supplement
the parameters you expected
\end{Description}
%
\begin{Usage}
\begin{verbatim}
param_primer3(x, add)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] is the Priemr3 input data

\item[\code{add}] supplementary parameters for Primer3
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
chromosome <- BSgenome.Hsapiens.UCSC.hg19
entend.hot<-Hot_extension(hot,185,chromosome)
p.hot<-Primer3_Hotset(entend.hot,return=2)
View(p.hot)
add<-c("PRIMER_INTERNAL_MAX_TM=63","PRIMER_MAX_END_GC=5")
param_primer3(p.hot,add)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Primer3\_Exoset}{Primer3\_Exoset}{Primer3.Rul.Exoset}
%
\begin{Description}\relax
'Primer3\_Hotset is used to form the file, which is required by Primer3. Here tarquire sequence is
Exome sequence, which has been sheared
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Primer3_Exoset(x, EX, generic = "generic", oligo = 0, rangel = 100,
  ranger = 150, opt_size = 20, min_size = 16, max_size = 22,
  explain_fiag = 1, return = 50, min_tm = 45, opt_tm = 60,
  max_tm = 60)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] is the is Exome dataframe after extension，here, each taget sequence requires the names (NO.)

\item[\code{EX}] total lenght of target sequence after amplifying, which is the same of n at 'Exo\_extension' function

\item[\code{generic}] is the input of The PRIMER\_TASK that tells primer3 which type of primers to pick.
You can select typical primers for PCR detection, primers for cloning or for sequencing. Here,
PRIMER\_TASK alsoccalled 'pick\_detection\_primers' is default setting  to 'generic' while
retaining 'pick\_detection\_primers' as an alias for backward compatibility.

\item[\code{oligo}] is the input of PRIMER\_PICK\_INTERNAL\_OLIGO, which default is 0 .If the associated
value = 1 (non-0), then primer3 will attempt to pick an internal oligo (hybridization probe to
detect the PCR product).

\item[\code{rangel}] is the left range  of PRIMER\_PRODUCT\_SIZE\_RANGE input. The associated values specify the lengths
of the product that the user wants the primers to create, and is a space separated list of elements
of the form <x>-<y>,where an <x>-<y> pair is a legal range of lengths for the product. For example,
if one wants PCR products to be between 100 to 150 bases (inclusive) then one would set this parameter
to 100-150. Here, default is 100-150

\item[\code{ranger}] is the lright  range  of PRIMER\_PRODUCT\_SIZE\_RANGE input.

\item[\code{opt\_size}] is the input of PRIMER\_OPT\_SIZE, which is the Optimum length (in bases) of a primer.
Primer3 will attempt to pick primers close to this length. Here, default is 20.

\item[\code{min\_size}] is the input of PRIMER\_MIN\_SIZE. The minimum acceptable length of a primer.
Must be greater than 0 and less than or equal to PRIMER\_MAX\_SIZE. Here,default is 16

\item[\code{max\_size}] is the input of PRIMER\_MAX\_SIZE, The maximum acceptable length of a primer.
Must be greater than 0 and less than or equal to PRIMER\_MIN\_SIZE. Here,default is 22

\item[\code{explain\_fiag}] is the input of If this flag is 1 (non-0), produce PRIMER\_LEFT\_EXPLAIN,
PRIMER\_RIGHT\_EXPLAIN, PRIMER\_INTERNAL\_EXPLAIN and/or PRIMER\_PAIR\_EXPLAIN output tags
as appropriate. These output tags are intended to provide information on the number
of oligos and primer pairs that primer3 examined and counts of the number discarded
for various reasons. If -format\_output is set similar information is produced in the
user-oriented output. Here,default is 1

\item[\code{return}] is the input of PRIMER\_NUM\_RETURN, which is the maximum number of primer (pairs) to return. Primer pairs
returned are sorted by their "quality", in other words by the value of the objective function
(where a lower number indicates a better primer pair). Caution: setting this parameter to a
large value will increase running time. Here,default is 5

\item[\code{min\_tm}] Minimum acceptable melting temperature (Celsius) for a primer oligo. Here,default is 45

\item[\code{opt\_tm}] Optimum melting temperature (Celsius) for a primer. Primer3 will try
to pick primers with melting temperatures are close to this temperature. The oligo
melting temperature formula used can be specified by user. Please see PRIMER\_TM\_FORMULA
for more information. Here,default is 60

\item[\code{max\_tm}] Maximum acceptable melting temperature (Celsius) for a primer oligo.Here,default is 60
\end{ldescription}
\end{Arguments}
%
\begin{Value}
dataframe
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
chromosome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
entend.exome<-Exo_extension(class$EXOME,280,200,chromosome)
p.exo<-Primer3_Exoset(exome.ex,EX=250,return=2)
View(p.exo)
write.table(exo,"exo.txt",quote = FALSE,row.names = FALSE, col.names = FALSE)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Primer3\_Hotset}{Primer3\_Hotset}{Primer3.Rul.Hotset}
%
\begin{Description}\relax
Primer3\_Hotset is used to form the file, which is required by Primer3. Here tarquire sequence
do not need tp be sheared
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Primer3_Hotset(x, generic = "generic", oligo = 0, rangel = 100,
  ranger = 150, opt_size = 20, min_size = 16, max_size = 22,
  explain_fiag = 1, return = 50, min_tm = 45, opt_tm = 60,
  max_tm = 60)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] is the is dataframe after extension，here, each taget sequence requires the names (NO.)

\item[\code{generic}] is the input of The PRIMER\_TASK that tells primer3 which type of primers to pick.
You can select typical primers for PCR detection, primers for cloning or for sequencing. Here,
PRIMER\_TASK alsoccalled 'pick\_detection\_primers' is default setting  to 'generic' while
retaining 'pick\_detection\_primers' as an alias for backward compatibility.

\item[\code{oligo}] is the input of PRIMER\_PICK\_INTERNAL\_OLIGO, which default is 0 .If the associated
value = 1 (non-0), then primer3 will attempt to pick an internal oligo (hybridization probe to
detect the PCR product).

\item[\code{rangel}] is the left range  of PRIMER\_PRODUCT\_SIZE\_RANGE input. The associated values specify the lengths
of the product that the user wants the primers to create, and is a space separated list of elements
of the form <x>-<y>,where an <x>-<y> pair is a legal range of lengths for the product. For example,
if one wants PCR products to be between 100 to 150 bases (inclusive) then one would set this parameter
to 100-150. Here, default is 100-150

\item[\code{ranger}] is the lright  range  of PRIMER\_PRODUCT\_SIZE\_RANGE input.

\item[\code{opt\_size}] is the input of PRIMER\_OPT\_SIZE, which is the Optimum length (in bases) of a primer.
Primer3 will attempt to pick primers close to this length. Here, default is 20.

\item[\code{min\_size}] is the input of PRIMER\_MIN\_SIZE. The minimum acceptable length of a primer.
Must be greater than 0 and less than or equal to PRIMER\_MAX\_SIZE. Here,default is 16

\item[\code{max\_size}] is the input of PRIMER\_MAX\_SIZE, The maximum acceptable length of a primer.
Must be greater than 0 and less than or equal to PRIMER\_MIN\_SIZE. Here,default is 22

\item[\code{explain\_fiag}] is the input of If this flag is 1 (non-0), produce PRIMER\_LEFT\_EXPLAIN,
PRIMER\_RIGHT\_EXPLAIN, PRIMER\_INTERNAL\_EXPLAIN and/or PRIMER\_PAIR\_EXPLAIN output tags
as appropriate. These output tags are intended to provide information on the number
of oligos and primer pairs that primer3 examined and counts of the number discarded
for various reasons. If -format\_output is set similar information is produced in the
user-oriented output. Here,default is 1

\item[\code{return}] is the input of PRIMER\_NUM\_RETURN, which is the maximum number of primer (pairs) to return. Primer pairs
returned are sorted by their "quality", in other words by the value of the objective function
(where a lower number indicates a better primer pair). Caution: setting this parameter to a
large value will increase running time. Here,default is 5

\item[\code{min\_tm}] Minimum acceptable melting temperature (Celsius) for a primer oligo. Here,default is 45

\item[\code{opt\_tm}] Optimum melting temperature (Celsius) for a primer. Primer3 will try
to pick primers with melting temperatures are close to this temperature. The oligo
melting temperature formula used can be specified by user. Please see PRIMER\_TM\_FORMULA
for more information. Here,default is 60

\item[\code{max\_tm}] Maximum acceptable melting temperature (Celsius) for a primer oligo.Here,default is 60
\end{ldescription}
\end{Arguments}
%
\begin{Value}
dataframe
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
chromosome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
Hot_extension(hot,185,chromosome)
p.hot<-Primer3_Hotset(entend.hot,return=2)
View(p.hot)
write.table(x,"p.hot.txt",quote = FALSE,row.names = FALSE, col.names = FALSE)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Split\_exome}{Split\_exome}{Split.Rul.exome}
%
\begin{Description}\relax
Split\_exome
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Split_exome(exo.out, frame)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{exo.out}] is the output exome result of Primer3

\item[\code{frame}] is the exome targe sequence dataframe
\end{ldescription}
\end{Arguments}
%
\begin{Value}
list
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
library(xxx)
data(exo.out)
data(exo)
sp.ex<-Split_exome(exo.out,frame=exo)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{TargetSet\_primer}{TargetSet\_primer}{TargetSet.Rul.primer}
%
\begin{Description}\relax
TargetSet\_primer
\end{Description}
%
\begin{Usage}
\begin{verbatim}
TargetSet_primer(x)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] List set of Primers under each target sequence converted by 'List\_PrimerSet' function
\end{ldescription}
\end{Arguments}
%
\begin{Value}
dataframe
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
plist<-List_PrimerSet(hot.out)
TargetSet_primer(plist)

exome
sp.ex<-Split_exome(exo.out,frame=exo)
eplist<-List_PrimerExomeSet(sp.ex)
lapply(eplist,TargetSet_primer) %>% list.rbind()
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Target\_primer}{Target\_primer}{Target.Rul.primer}
%
\begin{Description}\relax
This fucntion is used to get information of each primer under the target sequecne
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Target_primer(x)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{x}] List of Primers under each target sequence converted by 'List\_PrimerSet' function
\end{ldescription}
\end{Arguments}
%
\begin{Value}
data.frame
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
plist<-List_PrimerSet(hot.out)
Target_primer(plist[[1]])
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{visualGenome}{visualGenome}{visualGenome}
%
\begin{Description}\relax
visualGenome
\end{Description}
%
\begin{Usage}
\begin{verbatim}
visualGenome(Chr, Target.Start, Target.End, Extend.length, ID, x,
  plot = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{Chr}] Chromosome information，e.g "Chr1"

\item[\code{Target.Start}] Starting location of target sequence

\item[\code{Target.End}] ending location of target sequence

\item[\code{Extend.length}] toal length of extended target sequence

\item[\code{ID}] Name or NO. of target sequence

\item[\code{x}] a pair primer 3 avliable from Hot\_output or Exome\_outpur function

\item[\code{plot}] plot the genomic location, defaulr is to ‘TRUE’
\end{ldescription}
\end{Arguments}
%
\begin{Value}
GRange of target and primer pair information
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
Chr<-"chr5"
Target.Start<-176939369
Target.End<-176939371
Extend.length<-185
ID<-"leukemia_36"
plist<- List_PrimerSet(hot.out)
hot.pair<-List_PairSet(plist)
rem_Pairs.inf<-Filter_Inf(hot.pair,plist,distance=20,overlap=10)
x<-Hot_output(rem_Pairs.inf,Write.out=FALSE)[3,]
visualGenome(Chr,Target.Start,Target.End,Extend.length,ID,x)
#only return local information
visualGenome(Chr,Target.Start,Target.End,Extend.length,ID,x,plot=FALSE)
\end{ExampleCode}
\end{Examples}
\printindex{}
\end{document}
